consensus/src/model/stores/reachability.rs

```diff
@@ -52,19 +54,122 @@ pub trait ReachabilityStore: ReachabilityStoreReader {
     fn append_child(&mut self, hash: Hash, child: Hash) -> Result<u64, StoreError>;
     fn insert_future_covering_item(&mut self, hash: Hash, fci: Hash, insertion_index: usize) -> Result<(), StoreError>;
     fn set_parent(&mut self, hash: Hash, new_parent: Hash) -> Result<(), StoreError>;
-    fn replace_child(&mut self, hash: Hash, replaced_index: usize, replace_with: &[Hash]) -> Result<(), StoreError>;
-    fn replace_future_covering_item(&mut self, hash: Hash, replaced_index: usize, replace_with: &[Hash]) -> Result<(), StoreError>;
+    fn replace_child(
+        &mut self,
+        hash: Hash,
+        replaced_hash: Hash,
+        replaced_index: usize,
+        replace_with: &[Hash],
+    ) -> Result<(), StoreError>;
+    fn replace_future_covering_item(
+        &mut self,
+        hash: Hash,
+        replaced_hash: Hash,
+        replaced_index: usize,
+        replace_with: &[Hash],
+    ) -> Result<(), StoreError>;
     fn delete(&mut self, hash: Hash) -> Result<(), StoreError>;
     fn get_height(&self, hash: Hash) -> Result<u64, StoreError>;
     fn set_reindex_root(&mut self, root: Hash) -> Result<(), StoreError>;
     fn get_reindex_root(&self) -> Result<Hash, StoreError>;
 }
 
+/// Ordered DB set (manages a set per entry with cache and ordering).
+/// Used both for the tree children set and for the future covering set (per block)
+#[derive(Clone)]
+struct DbReachabilitySet {
+    access: DbSetAccess<Hash, Hash>,
+    cache: Cache<Hash, BlockHashes>,
+}
+
+impl DbReachabilitySet {
+    fn new(set_access: DbSetAccess<Hash, Hash>, set_cache: Cache<Hash, BlockHashes>) -> Self {
+        Self { access: set_access, cache: set_cache }
+    }
+
+    fn append(&mut self, writer: impl DbWriter, hash: Hash, element: Hash) -> Result<(), StoreError> {
+        if let Some(mut entry) = self.cache.get(&hash) {
+            Arc::make_mut(&mut entry).push(element);
+            self.cache.insert(hash, entry);
+        }
+        self.access.write(writer, hash, element)?;
+        Ok(())
+    }
+
+    fn insert(&mut self, writer: impl DbWriter, hash: Hash, element: Hash, insertion_index: usize) -> Result<(), StoreError> {
+        if let Some(mut entry) = self.cache.get(&hash) {
+            Arc::make_mut(&mut entry).insert(insertion_index, element);
+            self.cache.insert(hash, entry);
+        }
+        self.access.write(writer, hash, element)?;
+        Ok(())
+    }
+
+    fn replace(
+        &mut self,
+        mut writer: impl DbWriter,
+        hash: Hash,
+        replaced_hash: Hash,
+        replaced_index: usize,
+        replace_with: &[Hash],
+    ) -> Result<(), StoreError> {
+        if let Some(mut entry) = self.cache.get(&hash) {
```

Comment:
Can't we get `replaced_hash` with `entry[replaced_index]`? (Applies to other places as well)

Reply:
Note that the cache entry can be missing but we still need to update the DB in line 124. The caller always tracks this hash anyway so it makes sense to pass it to avoid forcing a read